From: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Date: Wed, 20 Jun 2018 19:22:06 +0200
Subject: Don't allow ubuntu mode extension to update

Ensure that no update is proposed or loaded if sideloaded (always
prefer system version) on the ubuntu session.
We want to ensure that the default code running is going through
our QA and security team process than being loaded from a 3rd
party website.
Also, that will enable us to upload newer versions on GNOME
extension website while still letting older ubuntu release versions
running expected extension version.
Note that the patch set is simplified with a hardcoded extension list
compared to https://bugzilla.gnome.org/show_bug.cgi?id=789852
which tracked mode extensions.
The patch set is too large otherwise to keep it maintainable as a distro
patch.
Origin: ubuntu
Forwarded: https://bugzilla.gnome.org/show_bug.cgi?id=789852
---
 js/misc/desktop.js           | 13 +++++++++++++
 js/misc/extensionUtils.js    |  7 +++++++
 js/ui/extensionDownloader.js | 13 +++++++++++++
 3 files changed, 33 insertions(+)

Index: gnome-shell-ubuntu/js/misc/desktop.js
===================================================================
--- gnome-shell-ubuntu.orig/js/misc/desktop.js
+++ gnome-shell-ubuntu/js/misc/desktop.js
@@ -6,6 +6,8 @@ const GLib = imports.gi.GLib;
 // the env variable. It's safe to cache matching result
 let _currentDesktopsMatches = {};
 
+let ubuntuModeExtensions = ["ubuntu-dock@ubuntu.com", "ubuntu-appindicators@ubuntu.com", "desktop-icons@csoriano"]
+
 // is:
 // @name: desktop string you want to assert if it matches the current desktop env
 //
@@ -37,3 +39,14 @@ function is(name) {
     _currentDesktopsMatches[name] = false;
     return false;
 }
+
+// Compared to https://bugzilla.gnome.org/show_bug.cgi?id=789852,
+// we can't now easily without having a dependency on
+// main.session from misc packages. This prevents g-s-d-prefs from loading.
+// Harcode the list for the ubuntu case until now.
+function isSystemPinnedExtension(uuid) {
+    if (!is("ubuntu"))
+        return false;
+
+    return ubuntuModeExtensions.indexOf(uuid) !== -1;
+}
Index: gnome-shell-ubuntu/js/ui/extensionDownloader.js
===================================================================
--- gnome-shell-ubuntu.orig/js/ui/extensionDownloader.js
+++ gnome-shell-ubuntu/js/ui/extensionDownloader.js
@@ -3,6 +3,7 @@
 const { Clutter, Gio, GLib, GObject, Soup } = imports.gi;
 
 const Config = imports.misc.config;
+const Desktop = imports.misc.desktop;
 const Dialog = imports.ui.dialog;
 const ExtensionUtils = imports.misc.extensionUtils;
 const FileUtils = imports.misc.fileUtils;
@@ -20,6 +21,14 @@ function installExtension(uuid, invocati
     let params = { uuid: uuid,
                    shell_version: Config.PACKAGE_VERSION };
 
+    if (Desktop.isSystemPinnedExtension(uuid)) {
+        let title = _("Can't install “%s”:").format(uuid);
+        let msg = _("This is an extension enabled by your current mode, you can't install manually any update in that session.");
+        Main.notifyError(title, msg);
+        invocation.return_dbus_error('org.gnome.Shell.CantInstallError', msg);
+        return;
+    }
+
     let message = Soup.form_request_new_from_hash('GET', REPOSITORY_URL_INFO, params);
 
     _httpSession.queue_message(message, (session, message) => {
@@ -152,6 +161,9 @@ function updateExtension(uuid) {
 function checkForUpdates() {
     let metadatas = {};
     Main.extensionManager.getUuids().forEach(uuid => {
+        // don't updates out of repository mode extension
+        if (Desktop.isSystemPinnedExtension(uuid))
+            return;
         metadatas[uuid] = Main.extensionManager.extensions[uuid].metadata;
     });
 
Index: gnome-shell-ubuntu/js/ui/extensionSystem.js
===================================================================
--- gnome-shell-ubuntu.orig/js/ui/extensionSystem.js
+++ gnome-shell-ubuntu/js/ui/extensionSystem.js
@@ -3,6 +3,7 @@
 const { Gio, St } = imports.gi;
 const Signals = imports.signals;
 
+const Desktop = imports.misc.desktop;
 const ExtensionUtils = imports.misc.extensionUtils;
 const FileUtils = imports.misc.fileUtils;
 const Main = imports.ui.main;
@@ -467,6 +468,10 @@ var ExtensionManager = class {
             let type = dir.has_prefix(perUserDir)
                 ? ExtensionType.PER_USER
                 : ExtensionType.SYSTEM;
+            if (Desktop.isSystemPinnedExtension(uuid) && type === ExtensionType.PER_USER) {
+                log(`Found user extension ${uuid}, but not loading from ${dir.get_path()} directory as part of session mode.`);
+                return;
+            }
             try {
                 extension = this.createExtensionObject(uuid, dir, type);
             } catch (e) {
